### PR TITLE
Master groupby m2m boi

### PR DESCRIPTION
**Pull request by brboi**
PURPOSE
Allow to group records by many2many fields.
Use case: turn the user_id m2o field on project.task into a user_ids m2m field in order to assign multiple collaborators on a task

SPECIFICATION
Add many2many fields to the 'list of groupable fields'.
When grouping a set of records by a many2many, records linked to multiple records of the comodel will essentially be in multiple 'groups' (even though those are not really 'groups' anymore).
There will also be a "null" group, for records that have an empty value of the many2many, like "unassigned" records.

Example: grouping a project.task linked to 10 res.users by the user_ids many2many field, the project.task will be 'rendered/considered/displayed' 10 times, once per res.users.
Kind of like if we grouped the m2m relation table instead.

Additional view specific changes when grouping by a m2m field:

    listview
        editing an instance of a record also updates each other instance (if any)
            the same goes when scheduling an activity, etc.
        disable deletion
            the risk here is that the user might expected that this will 'unlike' the record from a specific record of the many2many field
    kanban
        disable drag&drop between groups
        disable deletion
    pivot/graph
        do nothing special for now (i.e. let the m2m grouping behave 'naturally', even if it means having incorrect aggregates)
        we are aware that grouping a reporting view (graph, pivot) will display erroneous/misleading aggregate
            we decided not to address this at this stage
                options:
                    disable aggregates or grouping entirely (restricting the feature)
                    adding some kind of 'warning'

     

TASK 2508608
ENTERPRISE PR odoo/enterprise#20192

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/legacy/js/views/basic/basic_model.js [hunk 1]
```diff
@@ -1189,6 +1189,7 @@ var BasicModel = AbstractModel.extend({
 
                             self.unfreezeOrder(record.id);
 
+                            self.updateSameResIdRecords(record, _changes);
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/basic/basic_model.js:1192]: rename that to smth more obviously related to m2m ?

File: addons/web/static/src/legacy/js/views/kanban/kanban_column.js [hunk 2]
```diff
@@ -77,7 +78,7 @@ var KanbanColumn = Widget.extend({
 
         this.record_options = _.clone(recordOptions);
 
-        if (options.grouped_by_m2o || options.grouped_by_date ) {
+        if (options.grouped_by_m2o || options.grouped_by_date || options.grouped_by_m2m) {
             // For many2one and datetime, a false value means that the field is not set.
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/kanban/kanban_column.js:82]: adapt comment

File: addons/web/static/src/legacy/js/views/kanban/kanban_view.js [hunk 3]
```diff
@@ -68,6 +68,7 @@ var KanbanView = BasicView.extend({
         this.rendererParams.record_options = {
             editable: activeActions.edit,
             deletable: activeActions.delete,
+            archivable: activeActions.edit,
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/kanban/kanban_view.js:71]: looks weird

File: addons/web/static/src/legacy/js/views/list/list_controller.js [hunk 4]
```diff
@@ -918,7 +934,27 @@ var ListController = BasicController.extend({
...
+                    return this.update({}, { reload: false }).then(() => {
+                        if (ev.data.onSuccess) {
+                            ev.data.onSuccess();
+                        }
+                    });
+                } else {
+                    if (ev.data.onSuccess) {
+                        ev.data.onSuccess(result);
+                    }
+                }
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/list/list_controller.js:956]: feels weird

File: addons/web/static/src/legacy/js/views/list/list_model.js [hunk 5]
```diff
@@ -73,12 +75,39 @@
...
+                        utils.traverse_records(listState, function (r) {
+                            if (r.res_id === data.id && !recordIds.includes(r.id)) {
+                                sameResIdrecords.push(r);
+                            }
+                        });
+                        sameResIdrecords.forEach((rec) => {
+                            const record = self.localData[rec.id];
+                            updateLocalRecord(record, data);
+                        });
+                    }
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/list/list_model.js:110]: factorize with updateSameResIdRecords() ?

File: addons/web/static/tests/helpers/mock_server.js [hunk 6]
```diff
@@ -770,33 +770,100 @@ export class MockServer {
...
-            groupBy.forEach((groupByField) => {
-                value = (value ? value + "," : value) + groupByField + "#";
-                const fieldName = groupByField.split(":")[0];
-                if (fields[fieldName].type === "date") {
-                    value += formatValue(groupByField, record[fieldName]);
-                } else {
-                    value += JSON.stringify(record[groupByField]);
+        // if many2many field type comes then add 1 group for each m2m record like if groupby is ["m2o", "m2m"]
+        // then it should generate like "m2o#1,m2m#1" "m2o#1,m2m#2", "m2o#2,m2m#1", "m2o#2,m2m#2" and so on...
+        const generateGroups = (records) => {
```
Comment (brboi) [inline addons/web/static/tests/helpers/mock_server.js:775]: simplification is probably possible here

File: addons/mail/models/mail_activity.py [hunk 7]
```diff
@@ -1046,14 +1046,18 @@ def _read_progress_bar(self, domain, group_by, progress_bar):
         """.format(
             fields=', '.join(select_terms),
             from_clause=from_clause,
+            tz1=tz,
+            tz2=tz,
+            tz3=tz,
```
Comment (brboi) [inline addons/mail/models/mail_activity.py:1051]:
```suggestion
            tz=tz,
```

**Review by brboi — COMMENTED** [review]
keep tracks through some comments

File: addons/web/static/src/legacy/js/views/basic/basic_model.js [hunk 8]
```diff
@@ -1192,11 +1198,20 @@ var BasicModel = AbstractModel.extend({
...
+                                    self._updateSameResIDRecords(record.id, (id) => {
+                                        Object.assign(self.localData[id].data, record.data);
+                                    });
+                                }
+                                resolve(changedFields);
                             } else {
                                 _.extend(record.data, _changes);
+                                if (record.parentID) {
+                                    self._updateSameResIDRecords(record.id, (id) => {
+                                        Object.assign(self.localData[id].data, _changes);
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/basic/basic_model.js:1212]: _changes => record.data

File: addons/web/static/src/legacy/js/views/basic/basic_controller.js [hunk 9]
```diff
@@ -845,20 +845,17 @@ var BasicController = AbstractController.extend(FieldManagerMixin, {
...
-            // no db_id given, so reload the main record
-            prom = this.reload({
-                fieldNames: data.fieldNames,
-                keepChanges: data.keepChanges || false,
-            });
+            // reload the relational field given its db_id
+            prom = this.model.reload(db_id).then(this._confirmSave.bind(this, db_id));
         }
-        prom.then(ev.data.onSuccess).guardedCatch(ev.data.onFailure);
+        prom.then(onSuccess).guardedCatch(onFailure);
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/basic/basic_controller.js:858]: recheck that

File: addons/web/static/src/legacy/js/views/kanban/kanban_column.js [hunk 10]
```diff
@@ -77,8 +78,8 @@ var KanbanColumn = Widget.extend({
 
         this.record_options = _.clone(recordOptions);
 
-        if (options.grouped_by_m2o || options.grouped_by_date ) {
-            // For many2one and datetime, a false value means that the field is not set.
+        if (options.grouped_by_m2o || options.grouped_by_date || options.grouped_by_m2m) {
+            // For many2x and datetime, a false value means that the field is not set.
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/kanban/kanban_column.js:82]: check that

File: addons/web/static/src/legacy/js/views/basic/basic_model.js [hunk 11]
```diff
@@ -5274,6 +5291,33 @@ var BasicModel = AbstractModel.extend({
...
+     * each one of them.
+     *
+     * @protected
+     * @param {string} recordID local resource
+     * @param {(recordID: string)=>void} updateFn called for each record datapoint
+     *  representing the same res_id record
+     */
+    _updateSameResIDRecords(recordID, updateFn) {
+        const { model, res_id } = this.localData[recordID];
+        const topmostParentID = (() => { // FIXME get the topmost grouped list instead
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/basic/basic_model.js:5307]: todo

File: addons/web/static/src/legacy/js/views/list/list_controller.js [hunk 12]
```diff
@@ -316,7 +316,7 @@ var ListController = BasicController.extend({
      */
     _confirmSave: function (id) {
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/list/list_controller.js:317]: change api

File: addons/web/static/src/legacy/js/views/basic/basic_controller.js [hunk 13]
```diff
@@ -845,20 +845,19 @@ var BasicController = AbstractController.extend(FieldManagerMixin, {
...
         } else {
             // no db_id given, so reload the main record
-            prom = this.reload({
-                fieldNames: data.fieldNames,
-                keepChanges: data.keepChanges || false,
-            });
+            prom = this.reload({ fieldNames, keepChanges });
         }
-        prom.then(ev.data.onSuccess).guardedCatch(ev.data.onFailure);
+        prom.then(onSuccess).guardedCatch(onFailure);
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/basic/basic_controller.js:860]: nicer diff

File: addons/web/static/src/legacy/js/views/basic/basic_controller.js [hunk 14]
```diff
@@ -845,20 +845,19 @@ var BasicController = AbstractController.extend(FieldManagerMixin, {
...
         ev.stopPropagation(); // prevent other controllers from handling this request
-        var data = ev && ev.data || {};
-        var handle = data.db_id;
-        var prom;
-        if (handle) {
+        const { db_id, fieldNames, keepChanges, onSuccess, onFailure } = ev.data;
+        const { isM2MGrouped } = this.model.get(this.handle);
+        let prom;
+        if (isM2MGrouped) {
+            prom = this.model.reload(this.handle).then(this._confirmSave.bind(this, this.handle));
```
Comment (brboi) [inline addons/web/static/src/legacy/js/views/basic/basic_controller.js:852]: move to list controller and remove diff here

**Review by kebeclibre — COMMENTED** [review]
ok for me.
Todo though:
- confirm tat a x2m in a form view cannot be grouped.
- confirm that calling updateDuplicateRecords doesn't make sense after an onchange

**Pull request closed without merging**